### PR TITLE
[Experiment] Hack a global fixpoint expansion in conversion.

### PR DESCRIPTION
**Pull request by ppedrot**
Quick and dirty experiment following the Coq call. Rather than implementing a full-blown refolding of fixpoint everywhere, we experiment with refolding when converting only.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by herbelin** [comment]
Looks nice. So, since `cbv` has its own structure, the same trick could be done in `Cbv.FIX`. For `Reductionops.whd_state`, that could be an opportunity to introduce a proper structure (together with a lazy substitution).

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
I wasn't at the call, what is this supposed to achieve?

**Comment by ppedrot** [comment]
At the call we discussed the *sea snake* idea of clever fixpoint refolding, but this was motivated by user-friendliness. This PR is rather an experiment to assess the potential performance gain of not handling bare fixpoints in conversion.

**Comment by ppedrot** [comment]
This PR makes Equations loop on a function ominously annotated with the comment "(** Now we define an obviously non-terminating function. *)"... There is probably some progress issue somewhere when bluntly replacing the fix body by a constant.

**Comment by ppedrot** [comment]
This doesn't look like a great idea in the end. It's slower on some devs, and there is some completeness issue it seems.

**Comment by herbelin** [comment]
>  It's slower on some devs, and there is some completeness issue it seems.

What does it imply about a possible systematic refolding in reduction strategies? That conversion is less efficient with refolded fixpoints? That unguarded fixpoints, or at the least the one in Equations, need non-refolding to prevent loops?

**Comment by ppedrot** [comment]
> That conversion is less efficient with refolded fixpoints?

Unclear to me. I think we need more experiments.

> That unguarded fixpoints, or at the least the one in Equations, need non-refolding to prevent loops?

Actually it's not an unguarded fixpoint, it's [a wf fixpoint guarded by the axiom that the total relation is well-founded](https://github.com/mattam82/Coq-Equations/blob/729693206a944461807a9f3f78398d27bffff1e4/examples/general_recursion.v#L53). So there is something else going on here.

**Comment by ppedrot** [comment]
Now with refolding on lazy reduction only: @coqbot bench

**Comment by ppedrot** [comment]
There are some huge variations in the per-line diff, but overall this looks like a neutral change.

**Comment by herbelin** [comment]
FWIW, the error in coq_library_undecidability is of the following form:
```
Require Import List.
Goal Forall (eq 0) (List.map id (repeat 0 1)).
apply Forall_cons; [reflexivity|].
(* before: Forall (eq 0) []
   after: Forall (eq 0)
   ((fix map (l : list nat) : list nat :=
       match l with
       | nil => nil
       | a :: t => id a :: map t
       end) (repeat 0 0))
*)
```

**Comment by herbelin** [comment]
For the record, I rebased the branch at https://github.com/herbelin/github-coq/pull/new/ppedrot-global-fix-conversion (I believe this could be useful for #19309 too, cc @Janno).

**Comment by Janno** [comment]
Thanks! The code is actually already in #19309 although I have extended it to handle mutual fixpoints. If there is interested I can try to find some time to extract those pieces (although they are not hard to find).

I am not sure if you were already around when @ppedrot and I talked about this MR here during CUDW so let me try to remember what transpired.I think the main issue that came up is that refolding without re-checking universes (i.e. what `cbn` and `simpl`) is not sound: https://github.com/coq/coq/issues/5645. So it's definitely not right to take the current approach and use it in the kernel. I am following `cbn` and `simpl` in ignoring that issue for now.

Another thing (which I only noticed last week) that is missing from both this MR here and my extended version is the ability to refold global fixpoints of the form `Definition x := let a := b in fix go .. := ..`. `cbn` and `simpl` manage to do this but I have not yet understood how that works. The support for this is a bit shaky but it seems some developments rely on it. It might be something that CClosure-level fixpoint refolding should also do, be it in `kred` or, eventually (and with universe checking?) in the kernel.

**Comment by herbelin** [comment]
> The code is actually already in https://github.com/coq/coq/pull/19309 although I have extended it to handle mutual fixpoints.

I realized that I misunderstood your commit named "Apply #18672". I thought it was the original commit for `lazy`, but it is an adaptation of it for `kred`. So, of course, you don't care about the rebase wrt `cClosure.ml`.

I'll look at your adaptation to mutual fixpoints so that maybe it can be applied to PMP's "hack" (this PR).

> I think the main issue that came up is that refolding without re-checking universes (i.e. what `cbn` and `simpl`) is not sound: https://github.com/coq/coq/issues/5645.

I was not aware of that, thanks for the pointer. It does not seem to be redhibitory though, as I guess we could still properly compute the constraints to add to make it sound.

**Comment by herbelin** [comment]
A quick report on continuing the experiment made in this PR:
- I could identify the impact of refolding in `lazy` for the "old" unification used by `apply`/`rewrite`
- the "old" unification relies on two reduction heuristics:
  - `reduce` forces the full reduction of the argument of a head `match`/`fix`/`proj` that would trigger a iota (calling `whd_betaiota_deltazeta_for_iota_state` that uses `lazy` for the full reduction of the argument)
  - `expand` unfolds the head constant of an expression (followed by a beta-iota)
- to avoid unfolding a `fix` not immediately reducing, one need to do the modifications:
  - in `reduce`: to support expressions of the form `constant-unfolding-to-fix` applied to`term-reducing-to-a-constructor` in addition to expressions of the form `match`/`fix`/`proj` applied to `term-reducing-to-a-constructor`
  - in `expand`: to split the unfolding of a `constant-unfolding-to-fix` into two steps with a boolean remembering if this constant is to unfold or already virtually unfolded (together with the following beta-iota); for the latter, we can use a custom strategy of `lazy` that unfolds an head constant-unfolding-to-fix if it triggers a iota step

I have a branch implementing these modifications and which requires only 2 extra `simpl` in the stdlib to pass.

Depending on what CI would say, if there are too many changes to do with this approach, an alternative approach will be to provide a strategy closer from what is done currently, that is, that anytime a constant-unfolding-to-a-fix is unfolded to trigger a iota in `reduce`, or unfolded and trigger a iota in `expand`, then *all* inner (deep) iota-redexes resulting of the unfolding of the constant are triggered in turn (e.g., `List.map id (repeat 0 1)` does not only reduce to `cons 0 (List.map (repeat 0 0))` but recursively to `cons 0 []`).

In short, the empirical study of `unification.ml` shows that the key `head` reduction strategies are not beta, iota, delta, fix, cofix, etc. but rather:
- delta of a non-fix constant followed by the beta-iota redexes that it generates (as in `id (fun x => x) a` -> `a`)
- delta of a fix constant followed by the beta-iota it generates, together with refolding (as in `List.map id (repeat 0 1)` -> `cons 0 (List.map id (repeat 0 0))`)
- delta of a fix constant followed by the beta-iota it generates, together with refolding and together with the further beta-iota that not refolding yet would generate (as in `List.map id (repeat 0 1)` -> `cons 0 []`)
